src/frontend/src/handler/create_sink.rs

```diff
@@ -197,27 +270,118 @@ pub async fn handle_create_sink(
                     parallelism: parallelism.get(),
                 });
-        (sink, graph)
+
+        (sink, graph, target_table_catalog)
     };
 
+    let mut target_table_replace_plan = None;
+    if let Some(table_catalog) = target_table_catalog {
+        if !table_catalog.incoming_sinks.is_empty() {
+            return Err(RwError::from(ErrorCode::BindError(
+                "Create sink into table with incoming sinks has not been implemented.".to_string(),
+            )));
+        }
+
+        // Retrieve the original table definition and parse it to AST.
+        let [mut definition]: [_; 1] = Parser::parse_sql(&table_catalog.definition)
+            .context("unable to parse original table definition")?
+            .try_into()
+            .unwrap();
+        let Statement::CreateTable { source_schema, .. } = &mut definition else {
+            panic!("unexpected statement: {:?}", definition);
+        };
+        let source_schema = source_schema
+            .clone()
+            .map(|source_schema| source_schema.into_v2_with_warning());
+
+        // Create handler args as if we're creating a new table with the altered definition.
+        let handler_args = HandlerArgs::new(session.clone(), &definition, Arc::from(""))?;
+        let col_id_gen = ColumnIdGenerator::new_alter(&table_catalog);
+        let Statement::CreateTable {
+            columns,
+            constraints,
+            source_watermarks,
+            append_only,
+            ..
+        } = definition
+        else {
+            panic!("unexpected statement type: {:?}", definition);
+        };
+
+        let (mut graph, mut table, source) = generate_stream_graph_for_table(
+            &session,
+            target_table.unwrap(),
+            &table_catalog,
+            source_schema,
+            handler_args,
+            col_id_gen,
+            columns,
+            constraints,
+            source_watermarks,
+            append_only,
+        )
+        .await?;
+
+        table.incoming_sinks = table_catalog.incoming_sinks.clone();
+
+        // for now we only support one incoming sink
+        assert!(table.incoming_sinks.is_empty());
+
+        for _ in 0..(table_catalog.incoming_sinks.len() + 1) {
+            for fragment in graph.fragments.values_mut() {
+                if let Some(node) = &mut fragment.node {
+                    insert_merger_to_union(node);
+                }
+            }
+        }
+
+        // Calculate the mapping from the original columns to the new columns.
+        let col_index_mapping = ColIndexMapping::new(
+            table_catalog
+                .columns()
+                .iter()
+                .map(|old_c| {
+                    table.columns.iter().position(|new_c| {
+                        new_c.get_column_desc().unwrap().column_id == old_c.column_id().get_id()
+                    })
+                })
+                .collect(),
+            table.columns.len(),
+        );
```

Comment:
Is this step necessary? As the `columns` are not changed, the mapping should always be an identical mapping. Can we simply fill `None` and let the meta service skip this?

Reply:
Oh, I find it already been changed to `Option`. Cool!

Reply:
Would you mind document it in `.proto` for when it's `None`?